### DISCUSSION TITLE
Lesson:10 why ourToken.balanceOf(address(ourToken)) or tokenBalance is coming "0" ?

### DISCUSSION THREAD

**Discussion by codeplay22** [Cyfrin/foundry-full-course-cu#1548] [Q&A]
in DeployOurToken.sol 
``` INITIAL_SUPPLY = 1000 ether``` and this is being called in **OurTokenTest.t.sol** as an **ourToken** and when i try to find the balance of it in **setUp** function in **(OurTokenTest.t.sol )** by ``` uint256 tokenBalance = ourToken.balanceOf(address(ourToken));
        console.log( "tokenBalance is ",tokenBalance);```  it gives "0 " as a result of tokenBalance. Isn't it should log this "1000000000000000000000"
This is my OurToken.sol
```// SPDX-License-Identifier: SEE LICENSE IN LICENSE
pragma solidity ^0.8.18;
import {ERC20} from "@openzeppelin/contracts/token/ERC20/ERC20.sol";

contract OurToken is ERC20{

    constructor(uint256 initialSupply) ERC20("OurToken","OT"){
        _mint(msg.sender,initialSupply);
    }    

}
```

This is DeployOurToken.s.sol
```// SPDX-License-Identifier: SEE LICENSE IN LICENSE
pragma solidity ^0.8.18;
import {Script} from "forge-std/Script.sol";
import{OurToken} from "../src/OurToken.sol";


contract DeployOurToken is Script {
    uint256 public constant INITIAL_SUPPLY = 1000 ether;
    function run() external returns(OurToken) {
        
        vm.startBroadcast();
        OurToken ot = new OurToken(INITIAL_SUPPLY);
        vm.stopBroadcast();
        return ot;
    }
}
```

And this is OurTokenTest.t.sol
```// SPDX-License-Identifier: SEE LICENSE IN LICENSE
pragma solidity ^0.8.18;

import {Test,console} from "forge-std/Test.sol";
import {DeployOurToken} from "../script/DeployOurToken.s.sol";
import {OurToken} from "../src/OurToken.sol";
import {StdCheats} from "forge-std/StdCheats.sol";


contract OurTokenTest is Test {
    OurToken public ourToken;
    DeployOurToken public deployer;

    address bob = makeAddr("bob");
    address alice = makeAddr("alice");

    uint256 public constant STARTING_BALANCE = 100 ether;
    function setUp( ) public {
        deployer = new DeployOurToken();
        ourToken = deployer.run();
        vm.prank(address(msg.sender));
        ourToken.transfer(bob,STARTING_BALANCE);
        console.log("bobs balance is",ourToken.balanceOf(bob));
        uint256 tokenBalance = ourToken.balanceOf(address(ourToken));
        console.log( "tokenBalance is ",tokenBalance);
    }
 ```
 
 This is what i am getting
 
 ```
 Logs:
  bobs balance is 100000000000000000000
  tokenBalance is  0
```

**Comment by EngrPips** [comment]
Hello @codeplay22 , what you are getting is correct i believe. The balance of `bob` is correct and then why are you checking the balance of the token contract on itself ? , or am i not seeing correctly ?

which address is `ourToken` ?

**Reply by codeplay22** [reply]
Hello @EngrPips i know ```bob's``` balance is correct i don't have any doubt on that part, however i was trying to find the balance of ```ot```
in  **DeployOurToken.s.sol**  i basically wanted to know that when this line code executes ``` OurToken ot = new OurToken(INITIAL_SUPPLY);``` does ```ot``` get's the balance equals to ```INITIAL_SUPPLY```?

**Reply by EngrPips** [reply]
well it is `OurTokenTest` contract that would get the balance and you have also transferred all the token to bob in this line of code `ourToken.transfer(bob,STARTING_BALANCE);` of your `OurTokenTest` test. so the balance of would still be zero. 

This line of code `uint256 tokenBalance = ourToken.balanceOf(address(ourToken));` is wrong and it should be the below 

```javascript
uint256 tokenBalance = ourToken.balanceOf(address(this));
```

instead of sending all of the `STARTING_BALANCE` to bob send 50 ether and the remaining 50 ether would still be on the `OurTokenTest` contract balance 

So the balance from the below code should return 50 ether 

```javascript
uint256 tokenBalance = ourToken.balanceOf(address(this));
```

**Reply by codeplay22** [reply]
no, i'm not talking about that. I am talking about the ```INITIAL_SUPPLY``` in **DeployOurToken.sol** in this line  ```OurToken ot = new OurToken(INITIAL_SUPPLY);```
i am expecting that ```ourToken``` gets the  ```INITIAL_SUPPLY``` which is **1000 ether** in our **test setUp function** when we write this line of code ```ourToken = deployer.run();```
So i want to access 1000 ether from this ```ourToken```  how can i do that?

**Reply by EngrPips** [reply]
Honestly i am really confused as to what you are intending to do, but what i know is the sender of the transaction is always the `msg.sender` and your token mint to the `msg.sender` as shown below 

```javascript
pragma solidity ^0.8.18;
import {ERC20} from "@openzeppelin/contracts/token/ERC20/ERC20.sol";

contract OurToken is ERC20{

    constructor(uint256 initialSupply) ERC20("OurToken","OT"){
        _mint(msg.sender,initialSupply);
    }    

}
```

when you use `startBroadcast()` cheat foundry uses the default sender to send the transaction so that should be the address the initial balance would be mapped to. my little brain is finding it hard to catch up lol. pardon me.

**Reply by codeplay22** [reply]
okay,
So what i meant is in my ```DeployOurToken.s.sol``` i have this line of code ```OurToken ot = new OurToken(INITIAL_SUPPLY);``` and the 
**INITIAL_SUPPLY** is **1000 ethers in DeployOurToken.s.sol only**  . So i wanted to check in my ```setUp``` function in **OurTokenTest.t.sol** that when this line of code executes ```  deployer = new DeployOurToken(); ourToken = deployer.run();``` does ```ourToken``` gets 1000 ethers or not? if yes how? and if not, how?
According to me ```ourToken``` in ```DeployOurToken.s.sol``` should get ```1000 ethers``` but when i try to log it it says ```0```. how is that possible?

**Reply by EngrPips** [reply]
Seem like i am even more confused  now lol, you want `ourToken` the erc20 token to have a balance of 100 ether on itself, I am not sure that is correct, it is just like saying you want to check if an $100 note have $100. the `ourToken` is the contract that would map people to balance to their address and whenever you deploy it , it would map the `1000 ethers` to whoever  deployed the contract not to itself. 

Long story short `OurToken` is the token contract so we check what the balance of every other address is on `OurToken`. you don't check the balance of `OurToken` on `OurToken`

```javascript
pragma solidity ^0.8.18;
import {ERC20} from "@openzeppelin/contracts/token/ERC20/ERC20.sol";

contract OurToken is ERC20{

    constructor(uint256 initialSupply) ERC20("OurToken","OT"){
        _mint(msg.sender,initialSupply);
    }    

}
```

You clearly instructed `OurToken` to mint the initialSupply to whoever deployed the `OurToken` contract and not to itself, maybe if you want to mint `OurToken` token to `OurToken` contract address you should change your code to the below and see if it works 

```javascript
pragma solidity ^0.8.18;
import {ERC20} from "@openzeppelin/contracts/token/ERC20/ERC20.sol";

contract OurToken is ERC20{

    constructor(uint256 initialSupply) ERC20("OurToken","OT"){
        _mint(address(this),initialSupply);
    }    

}
```
